Add annotations support for tf.estimator.Estimator

yongtang opened the pull request:
This fix adds annotations support for tf.estimator.Estimator so that the following works in python 3:
```
import tensorflow as tf

def model_fn(features: dict, labels: tf.Tensor, mode: str):
    pass

estimator = tf.estimator.Estimator(model_fn)
```

This fix fixes #12249.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

tensorflow-jenkins commented:
Can one of the admins verify this patch?

martinwicke commented on tensorflow/python/util/tf_inspect.py:
```diff
@@ -45,6 +45,27 @@ def getargspec(object):  # pylint: disable=redefined-builtin
+    will be called directly on the callable.
+  """
+  # The getfullargspec is called in estimator.Estimator (See issue 12249)
+  decorators, target = tf_decorator.unwrap(object)
```
If you unwrap it, it won't be the outermost object. I.e. if I have a decorator modifying the arguments, we should return the arguments directly, without unwrapping.

yongtang replied:
@martinwicke The unwrap was done previously so the implementation before PR might have already been problematic. Let me take a look and update the PR accordingly.

martinwicke commented on tensorflow/python/util/tf_inspect.py:
```diff
@@ -45,6 +45,27 @@ def getargspec(object):  # pylint: disable=redefined-builtin
                if d.decorator_argspec is not None), _inspect.getargspec(target))
 
 
+def getfullargspec(object):  # pylint: disable=redefined-builtin
```
Please name this something else, not "object", which is a reserved word.

martinwicke commented on tensorflow/python/util/tf_inspect.py:
```diff
@@ -45,6 +45,27 @@ def getargspec(object):  # pylint: disable=redefined-builtin
+    object: A callable, possibly decorated.
+
+  Returns:
+    The `FullArgSpec` (Python 3) or `ArgSpec` (Python 2)
```
Are these classes compatible enough that all downstream code can use them interchangeably? I.e. is FullArgSpec backwards compatible with ArgSpec?

yongtang replied:
@martinwicke They are not truly compatible as FullArgSpec has `varkw` while ArgSpec has `keywords`. For example, if `getfullargspec` is used, then in `tensorflow/python/util/deprecation.py`:

https://github.com/tensorflow/tensorflow/blob/325fc2330f4e527b394456ce1d821f2c73b1e7a2/tensorflow/python/util/deprecation.py#L224-L229

The following has to be used:
```
# Cannot use arg_spec.keywords directory
getattr(arg_spec, "keywords", getattr(arg_spec, "varkw"))
```

For that I try to use `util.getfullargspec` only for `estimator.Estimator`. For everything else, the original `util.getargspec` would be used.

yongtang committed:
Add annotations support for tf.estimator.Estimator
This fix adds annotations support for tf.estimator.Estimator
so that the following works in python 3:
```
import tensorflow as tf

def model_fn(features: dict, labels: tf.Tensor, mode: str):
    pass

estimator = tf.estimator.Estimator(model_fn)
```

This fix fixes 12249.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
@martinwicke I traced through the model_fn and the call with args. The model_fn is actually called with the args directly so as you pointed out the unwrapping may not be appropriate here.

I updated the PR and now the getfullargspec is called to the function directly.

Please take a look and let me know if there are any issues.

drpngx commented:
@martinwicke take another look?

sb2nov commented:
@martinwicke ping

martinwicke commented:
So this works. I'm sorry I wasn't clear before, it is good to use tf_inspect instead of inspect, so it can use TFDecorator decorated functions as model_fn. 

To do that, use tf_inspect. getargspec, and ideally make that function compatible with py3 by adding the switch for getfullargspec as you did inline.

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/util/tf_inspect.py#L32

Wee7 commented:
Hi, this also happens to me when I use tf.contrib.variance_scaling_initializer()

`initializer = tf.contrib.variance_scaling_initializer()

  File "/Users/WeeYin/anaconda/lib/python3.5/inspect.py", line 1050, in getargspec
    raise ValueError("Function has keyword-only arguments or annotations"

ValueError: Function has keyword-only arguments or annotations, use getfullargspec() API which can support them`

yongtang committed:
Update tf_inspect with getfullargspec

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
@Wee7 I am not able to reproduce with the current head:
```python
ubuntu@ubuntu:~$ python3
Python 3.5.2 (default, Nov 17 2016, 17:05:23) 
[GCC 5.4.0 20160609] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import tensorflow as tf
>>> initializer = tf.contrib.layers.variance_scaling_initializer()
```

```python
ubuntu@ubuntu:~$ python
Python 2.7.12 (default, Nov 19 2016, 06:48:10) 
[GCC 5.4.0 20160609] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import tensorflow as tf
>>> initializer = tf.contrib.layers.variance_scaling_initializer()
>>>
```

In case you are able to reproduce, could you open a separate issue and list the details steps to reach the error?

yongtang commented:
@martinwicke Thanks for the review. The PR has been updated. Please take a look.

martinwicke commented:
Sorry for the delay, can you update the PR and resolve the conflicts?

martinwicke commented:
Also, there's one unresolved comment (use of "object" as a arg name)

yongtang commented:
Thanks @martinwicke. The PR has been rebased and `object` has been changed to `obj` as well. Please take a look.

martinwicke commented:
Jenkins, test this please.

caisq commented:
@tensorflow-jenkins test this please

martinwicke commented:
ignoring stale Kokoro.
